src/include/duckdb/common/types/string.hpp

```diff
@@ -0,0 +1,270 @@
+		return !(*this < other);
+	}
+	bool operator<=(const String &other) const {
+		return !(*this > other);
+	}
+
+	char operator[](const idx_t pos) const {
+		D_ASSERT(pos < size());
+
+		if (IsInline()) {
+			return buf[pos];
+		}
+		return ptr[pos];
+	}
+
+public:
+	// STL-like interface
+	// NOLINTBEGIN - mimic std::string interface
+	size_t size() const {
+		return len & ~NON_OWN_BIT;
+	}
+	bool empty() const {
+		return len == 0;
+	}
+	const char *data() const {
+		return IsInline() ? buf : ptr;
+	}
+	const char *begin() const {
+		return data();
+	}
+	const char *end() const {
+		return data() + size();
+	}
+	const char *c_str() const {
+		return data();
+	}
+	// NOLINTEND
+
+	// Helper methods
+	bool IsOwning() const {
+		return (len & NON_OWN_BIT) == 0;
+	}
+	bool IsInline() const {
+		return len <= INLINE_MAX;
+	}
+	static bool CanBeInlined(size_t size) {
+		return size <= INLINE_MAX;
+	}
+
+	// Creates a new String instance with its own copy of the data
+	static String Copy(const char *data, size_t size) {
+		String result;
+		result.AssignOwning(data, size);
+		return result;
+	}
+
+	static String Copy(const String &other) {
+		return Copy(other.data(), other.size());
+	}
+	static String Copy(const char *data) {
+		return Copy(data, data ? strlen(data) : 0);
+	}
+	static String Copy(const std::string &str) {
+		return Copy(str.data(), str.size());
+	}
+	String Copy() const {
+		return Copy(data(), size());
+	}
+
+	// Creates a new String instance that references the data without owning it
+	// If the size is small enough, it will inline the data; which WILL be owning
+	static String Reference(const char *data, size_t size) {
+		String result;
+		// If we reference, and we can inline it, we make owning anyway
+		if (size <= INLINE_MAX) {
+			result.AssignOwning(data, size);
+		} else {
+			result.ptr = const_cast<char *>(data); // NOLINT allow const cast
+			result.len = size | NON_OWN_BIT;       // Set the non-owning bit
+		}
+		return result;
+	}
+
+	static String Reference(const String &other) {
+		return Reference(other.data(), other.size());
+	}
+	static String Reference(const char *data) {
+		return Reference(data, data ? strlen(data) : 0);
+	}
+	static String Reference(const std::string &str) {
+		return Reference(str.data(), str.size());
+	}
+	String Reference() const {
+		return Reference(data(), size());
+	}
+
+	string ToStdString() const {
+		if (IsInline()) {
+			return string(buf, size());
+		}
+		return string(ptr, size());
+	}
+
+public:
+	static char CharacterToLower(char c) {
+		if (c >= 'A' && c <= 'Z') {
+			return UnsafeNumericCast<char>(c + ('a' - 'A'));
+		}
+		return c;
+	}
+
+	String Lower() const {
+		const auto str_data = data();
+		const auto str_size = size();
+
+		std::string lowercase_str;
+		lowercase_str.reserve(str_size);
+		for (idx_t i = 0; i < str_size; ++i) {
+			lowercase_str.push_back(CharacterToLower(str_data[i]));
+		}
+		return String(lowercase_str);
```

Comment:
Connected to previous comment on allocations in `AssignOwning`.

For example, here calling `String(lowercase_str)` might trigger allocations, and there is really no way to preallocate this, it's this not a footgun?

Again, unsure how to solve this.

Reply:
I think the idea is that you can use `String` just like a normal `std::string`, i.e. assume that it's owning most of the time by default. if you construct, you get an owning string. You might allocate extra, but you wont segfault due to interface misuse.  

Then there is a `String::Reference(lower_case_str)` to _explicitly_ make a non-owning string. Referencing is explicit as you have to know what you're doing and have to ensure that the string is alive as long as the underlying data.

This is also why we don't have a copy constructor by default, you have to explicitly say if you want to `.Copy()` or `.Reference()`. But if we did have a copy constructor, it should probably also always create a owning copy, to avoid use-after-free.